packages/functional/src/pipe.ts

```diff
@@ -0,0 +1,39 @@
+import { BaseTransaction } from '@solana/transactions';
+
+type Func<T, R> = (a: T) => R;
+
+type Chain<T, R, F extends Func<T, R>[]> = F extends []
+    ? T
+    : F extends [Func<infer U, infer V>, ...infer Rest]
+    ? T extends U
+        ? V & Chain<T, R, Rest extends Func<T, R>[] ? Rest : []>
+        : never
+    : never;
+
+/**
+ * General pipe function.
+ * Provide an initial value and a list of functions to pipe it through.
+ * @param a     The initial value
+ * @param fns   Any number of functions to pipe the value through
+ * @returns     The final value with all functions applied
+ */
+export function pipe<T, R, F extends Func<T, R>[]>(a: T, ...fns: F) {
+    return fns.reduce((acc, fn) => {
+        return { ...acc, ...fn(acc) };
+    }, a as Chain<T, R, F>);
+}
+
+/**
+ * Pipe for creating transactions.
+ * This function drives the general `pipe` function, but provides type safety for
+ * working with transactions.
+ * @param baseTransaction   The initial transaction, which can be any type of transaction
+ * @param mutations         Any number of mutating functions to pipe the transaction through
+ * @returns                 The final transaction with all mutations applied
+ */
+export function pipeTransaction<T extends BaseTransaction, R extends BaseTransaction, F extends Func<T, R>[]>(
```

Comment:
We shouldn't need this.

```ts
const tx = pipe(
   createTransaction({ version: 0 }),
   () => { return { joe: 'string' } },
   setFeePayer,
   setBlockhash,
)
```

That generic invocation alone should be able to know, statically, that `{joe: 'string'}` does not conform to any of the input types _demanded_ by `setTransactionFeePayer`.

But you know what _can_ know that, statically?

https://www.typescriptlang.org/play?ssl=10&ssc=1&pln=10&pc=2#code/KYDwDg9gTgLgBAMwK4DsDGMCWEVzJsYAHgBUBJFTLAQwBsA+ACk0pgC45zXM6BKDrlR60AUKEixEqDNlz5CpCkLoAaOACUAjExZUBSmrTW6YAfSiaOzVvu584AXnobN-F2PDR4ydFhx4CYkFDNS1QgCYdG04DYWNWc0s4az0Yu1peR2cwuAtzcKsLDi1Mpw1wt3VwjwlvaT85QMV0qNTguLgAOm6EFABnDgBhOlpqACNaYAAxetkAbQBdTIBvOABfERF5YEYROH24Oc01cLUAZgWVPYPGUudGVYgxgCtgDA4AchQIAHcP9d4VwOyWoUCgHBQSAAtmNgFBFnc4KCoJ0odQwIxcGVcABqOCuES8IA

Your old friend… :)

Reply:
![sob](https://media.giphy.com/media/WHamTKpzAMeiR0k8Nf/giphy.gif)

That's hilarious. I guess I'll have to add those and we'll have to have a max cap eh?

Reply:
Does this work to combine objects? I see it does change the type but one problem I faced was when the first couple of functions remove signatures, then the others add them, and there's also multiple signatures being appended on each other.

Rhetorical, I can mess around with it.